chain/tests/test_force_upgrade.rs

```diff
@@ -0,0 +1,114 @@
+//
+//     let statedb = chain.get_state_view();
+//     let account = Account::new_association();
+//     let sequence_number = statedb.get_sequence_number(account.address().clone())?;
+//
+//     let signed_txns = ForceUpgrade::begin(
+//         account,
+//         sequence_number,
+//         chain.info().chain_id(),
+//         FORCE_UPGRADE_BLOCK_NUMBER,
+//         statedb,
+//         statedb,
+//     )?;
+//
+//     let txns: Vec<Transaction> = signed_txns
+//         .iter()
+//         .cloned()
+//         .map(Transaction::UserTransaction)
+//         .collect();
+//
+//     let txn_outupts = starcoin_executor::execute_transactions(&statedb, txns.clone(), None)?;
+//     assert!(
+//         !txns.is_empty() || !txn_outupts.is_empty(),
+//         "Failed to execution"
+//     );
+//     let txn_output = txn_outupts.get(0).unwrap();
+//     assert_eq!(
+//         txn_output.status(),
+//         &TransactionStatus::Keep(KeptVMStatus::Executed),
+//         "Execute the deploy failed"
+//     );
+//     assert!(
+//         !txn_output.write_set().is_empty(),
+//         "Execute the deploy failed"
+//     );
+//
+//     ForceUpgrade::finish(statedb)?;
+//
+//     Ok(())
+// }
+
+#[stest::test]
+pub fn test_force_upgrade_in_openblock() -> anyhow::Result<()> {
+    let config = Arc::new(NodeConfig::random_for_test());
+    let chain = test_helper::gen_blockchain_for_test(config.net())?;
+    let header = chain.current_header().clone();
+
+    let block_gas_limit = 10000000;
+
+    let mut opened_block = {
+        let miner_account = AccountInfo::random();
+        OpenedBlock::new(
+            chain.get_storage(),
+            header,
+            block_gas_limit,
+            miner_account.address,
+            config.net().time_service().now_millis(),
+            vec![],
+            U256::from(0),
+            chain.consensus(),
+            None,
+        )?
+    };
+
+    let account_reader = chain.chain_state_reader();
+    let association_sequence_num =
+        account_reader.get_sequence_number(account_config::association_address())?;
+    let (_receive_prikey, receive_public_key) = KeyGen::from_os_rng().generate_keypair();
+    let receiver = account_address::from_public_key(&receive_public_key);
+    let txn1 = build_transfer_from_association(
+        receiver,
+        association_sequence_num,
+        50_000_000,
+        config.net().time_service().now_secs() + DEFAULT_EXPIRATION_TIME,
+        config.net(),
+    )
+    .try_into()?;
+
+    let excluded = opened_block.push_txns(vec![txn1])?;
+    assert_eq!(excluded.discarded_txns.len(), 0);
+    assert_eq!(excluded.untouched_txns.len(), 0);
+
+    let txns = {
+        let account = Account::new_association();
+        ForceUpgrade::force_deploy_txn(
+            account,
+            association_sequence_num + 1,
+            &opened_block.chain_id(),
+        )?
+    };
+    if !txns.is_empty() {
```

Comment:
This test is not compiling, and force_deploy txn will be discarded.

Reply:
fixed